### ISSUE TITLE
Infinite recursion fetching origin policy

### ISSUE THREAD

**Issue by domenic** [mikewest/origin-policy#84]
I _believe_ that the current text at https://wicg.github.io/origin-policy/#monkeypatch-fetch causes infinite recursion, because it performs a fetch, which will fetch the origin policy, which performs a fetch, which will fetch the origin policy...

I guess we just want to add another field to request which is something like "skip origin policy".

**Comment by domenic** [comment]
@annevk I'll probably be working on this tomorrow; if you have any thoughts about better architectures than just adding another request flag, let me know. Otherwise, we can always refactor later.

**Comment by annevk** [comment]
I guess I would expect something analogous to CORS preflights, but in part this depends how deeply we want to intertwine things. There's also a question for which URLs we should fetch the origin policy I suppose. Should we fetch it when fetching a blob URL?

**Comment by domenic** [comment]
> I guess I would expect something analogous to CORS preflights

It looks like the technique used there is to have the fetch be a HTTP-network-or-cache fetch instead of a main fetch. I guess that would work here too, if we continued to perform the origin policy processing in main fetch ([see current spec](https://wicg.github.io/origin-policy/#monkeypatch-fetch)). That does seem a bit nicer, and I went through all the differences between main fetch and HTTP-network-or-cache fetch and they seem to work as we want:

- Skipping upgrade-insecure-requests, HSTS, FTP blocking, bad port blocking, mixed content checks, CORS handling, and blocking certain opaque Range responses don't have any impact since the origin is the same and this is HTTPS only.
- Referrer policy inheritance doesn't matter since we're setting no-referrer
- MIME type and nosniff handling doesn't matter; we do our own MIME checking
- We don't have any integrity metadata to check
- Service workers don't matter because we're skipping them anyway
- Redirect handling doesn't matter since we're disallowing them anyway

The only one I'm unsure about is that CSP does not apply to HTTP-network-or-cache fetch, but that's probably reasonable.

> There's also a question for which URLs we should fetch the origin policy I suppose. Should we fetch it when fetching a blob URL?

Hmm. Currently the algorithms seem to fetch it for blob: URLs; nothing stops them. I'm not sure what the latest plan is for blob: URL inheriting things... If it inherits all the policies that could be set via origin policy anyway, then that seems a bit redundant. But probably not harmful?
